native/protocol/src/x3dh.rs

```diff
@@ -0,0 +1,380 @@
+    // After sending this, Alice may continue using SK or keys derived from SK within the post-X3DH protocol for communication with Bob
+    let ciphertext = encrypt_data(
+        Payload {
+            msg: message,
+            aad: &associated_data,
+        },
+        &ChaCha20Poly1305::new_from_slice(&sk).unwrap(),
+    )?;
+
+    Ok((
+        sk,
+        Message {
+            sender_identity,
+            sender_ik: sender_ik.verifying_key(),
+            ek,
+            opk: prekey_bundle.opk,
+            ciphertext,
+        },
+    ))
+}
+
+fn initiate_recv_get_sk(
+    sender_ik: &VerifyingKey,
+    ek: X25519PublicKey,
+    opk: Option<X25519StaticSecret>,
+    receiver_ik: &SigningKey,
+    spk: &X25519StaticSecret,
+) -> [u8; 32] {
+    let dh1 = spk.diffie_hellman(&X25519PublicKey::from(sender_ik.to_montgomery().to_bytes()));
+    let dh2 = X25519StaticSecret::from(receiver_ik.to_scalar_bytes()).diffie_hellman(&ek);
+    let dh3 = spk.diffie_hellman(&ek);
+
+    if let Some(opk) = opk {
+        // Bob deletes any one-time prekey private key that was used, for forward secrecy.
+        let dh4 = opk.diffie_hellman(&ek);
+        kdf(&[
+            dh1.to_bytes(),
+            dh2.to_bytes(),
+            dh3.to_bytes(),
+            dh4.to_bytes(),
+        ]
+        .concat())
+    } else {
+        kdf(&[dh1.to_bytes(), dh2.to_bytes(), dh3.to_bytes()].concat())
+    }
+}
+
+// Caller must delete sk on error.
+// opk must be wiped.
+pub fn initiate_recv(
+    receiver_ik: &SigningKey,
+    receiver_spk: &X25519StaticSecret,
+    sender_ik: &VerifyingKey,
+    ek: X25519PublicKey,
+    receiver_opk: Option<X25519StaticSecret>,
+    ciphertext: &[u8],
+) -> Result<([u8; 32], Vec<u8>), X3DHError> {
+    // Upon receiving Alice's initial message, Bob retrieves Alice's identity key and ephemeral key from the message.
+    // Bob also loads his identity private key, and the private key(s) corresponding to whichever signed prekey and one-time prekey (if any) Alice used.
+    // Using these keys, Bob repeats the DH and KDF calculations from the previous section to derive SK, and then deletes the DH values.
+    let sk = initiate_recv_get_sk(sender_ik, ek, receiver_opk, receiver_ik, receiver_spk);
+
+    // Bob then constructs the AD byte sequence using IKA and IKB, as described in the previous section.
+    // AD = Encode(IKA) || Encode(IKB)
+    let ad = [sender_ik.to_bytes(), receiver_ik.verifying_key().to_bytes()].concat();
+
+    // Bob may then continue using SK or keys derived from SK within the post-X3DH protocol for communication with Alice.
+    // Finally, Bob attempts to decrypt the initial ciphertext using SK and AD.
+    let cipher = ChaCha20Poly1305::new_from_slice(&sk).unwrap();
+    Ok((sk, decrypt_data(&ciphertext, &ad, &cipher)?))
+}
+
+#[cfg(test)]
```

Comment:
Future test enhancements I'd love to see:
1.  Add [Duvet](https://github.com/awslabs/duvet) to ensure you implement and have a test for every requirement from the standard. e.g. s2n-quic's [report](https://dnglbrstg7yg.cloudfront.net/latest/compliance.html#/)
2. Interoperability tests with another independent X3DH implementation